Documentation/Coherency.md

```diff
@@ -0,0 +1,74 @@
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the build.
+- Probably does not satisfy requirements for no prebuilt binaries.
+
+# Proposed Solution #
+---------------------
+Hybrid approach: we will enable storing and retrieving packages, but allow building multiple versions of each repo completely from source.  Daily builds and dev builds will mostly use the stored packages approach while release builds will use the completely-built-from-source approach.  Doing multiple versions of repos only for release builds should cut down on the numbers of builds that are done many times since these builds should be mostly coherent except for servicing considerations.
+
+## Phase 1 ##
+After Darc merges uberclone support (https://github.com/dotnet/arcade-services/pull/291), update our version of Darc and get rid of submodules, using an uberclone instead.  This will give us many more repos than we will actually use.  For the time being, use some `PreferredRepoHashes.props` or similar to tell the build which one to use - this should change nothing in our output.  At this point we should be able to take daily build updates but will be producing a product that differs from the official release, and will have problems if we don't implement later phases before servicing.
+
+## Phase 2 ##
+Implement some parts of repo isolation (https://github.com/dotnet/source-build/issues/928).  Use a separate output feed for each repo, then supply only the feeds that repo.projs say they depend on to each repo build.  Next implement publishing these feeds to our package storage.  It's nice to do this even if we don't have consumption yet because we will build up a cache of source-built packages we can use later.
+
+## Phase 3 ##
+Change to using each repo's `Version.Details.xml` to determine its inputs.  This includes the hash and the exact packages and version numbers of the repos, so after we have this we can get rid of `PreferredRepoHashes.props` and flip the switch to allow multiple builds per repo.  We will also be building the same bits as the official release.  We will also have to be very far along on patch removal at this point, or we'll run into lots of problems trying to keep multiple versions of patches straight.
+
+## Phase 4 ##
+Allow consuming packages from the package storage.  This can either be a master switch or a per-repo determination to allow easier incremental builds.
+
+## Mop-up ##
+Implement the rest of repo isolation.  `Versions.Details.xml` actually specifies exact packages so we can also supply only what a repo needs on a package-by-package basis.  Implement per-submodule source-build legs to increase our resilience to breaking changes.
+
+# Appendix #
+------------
+## Storage format proposal ##
+- dotnetcli source-build container
+- make a new source-built-packages folder
+- under this, use the format {repo}/{hash}/{package}
```

Comment:
In official builds, the same commit hash may be built multiple times resulting in different versions, but those would collide here. I think this matters for us because we need to match the official build versions in source-build releases.

Reply:
True, I wasn't considering that even version numbers can be incoherent (I was just thinking we could look them up from BAR).

Reply:
Version numbers of assets can be looked up in the BAR, but yeah you can build a repo multiple times.  The dependency graph already knows about this. Based on the dependencies listed in the Version.Details.xml, it will try to figure out which builds could contribute.  It could be multiple, but typically is 1.  When we get to servicing and stabilized builds, it will get more complicated, as multiple input builds could produce the exact same package versions. In that case, latest will be the right way to go.

Reply:
That's interesting, hadn't actually thought about how even commit + version isn't necessarily a unique build. OTOH, with ProdCon v2 I don't think we need to care because dependencies are tied to the commit and AFAIK there's no overarching thing to pass in queue-time params to change build behavior that way. (Contrasting with having to care with ProdCon v1 because of the PVP overrides and queue-time params.)

Reply:
We just need to make sure, if caching, that we are passing in the feed with the right prebuilts.